docs/request-for-comment/2022-05-11-microservice-language-selection.md

```diff
@@ -0,0 +1,29 @@
+### Discussed in
+
+<sup>Originally posted</sup>
+- Start  2022-03-24
+- Related Issue: https://vfs.atlassian.net/wiki/spaces/IET/pages/2114191369/RFC+File+Upload+will+be+developed+in+python
+
+# Summary
+
+`vets-api` is implemented as a monolithic ruby on rails application. IET is responsible for exploring how this monolith may be broken into independent services. This presents an opportunity for teams to leverage the test best suited for the microservice. The vanguard of microservice development, IET intends to develop file upload in python. 
+
+# Motivation
+
+A core objective of a shift from monolithic architecture to microservices is to shift pieces of functionality into independently deployable, loosely coupled services. Encapsulating functionality into discrete services reduces the overall complexity of the code that each team is responsible for, which makes the overall code base more testable and maintainable. Each microservice should be capable of operating independently, which introduces the possibility of developing services in languages other than ruby on rails. As teams split services from `vets-api` they are no longer required to develop these services using ruby on rails. 
+
+IET proposes that teams developing microservices should select a language based on which technical stack they believe best suits the needs of their service. File upload is a low-risk opportunity to demonstrate the feasibility of removing the requirement that all platform services be built in Ruby on Rails. The service is self-contained and technically straightforward, which will allow the team to more easily assume the risks that are associated with introducing a new language to the platform. 
+
+Regardless of which language the individual microservices are written in, in order for `vets-api` to successfully shift to a microservices architecture and sustain these services in production, frameworks for deploying, monitoring, discovering and testing the various services – independent of the language they are written in – will need to be identified and developed. Introducing the requirement that these frameworks function regardless of the language they are implemented in sets an important precedent for system design.
+
```

Comment:
Great points here, and definitely things that should be answered before we launch into a new microservice. As you mentioned, some of these are probably not questions that my team alone will solve, but here are my general thoughts. 

**Who will maintain the service after the contract for this team ends?**

This is definitely something that I think we should define beyond the implicit assumption that all platform solutions are supported by the platform team in the long term. Part of the service design for microservices should include plans for transitioning services from spike to platform. This is a general problem that will need to be solved regardless of the language that we select. 

One of the charters of IET is to determine the feasibility of microservices as a way of simplifying `vets-api` — I think that there is value in tackling the risks of this paradigm shift early. Microservices as a general concept may compound the existing problem of shepherding a particular piece of functionality throughout its lifecycle. 

**How will the microservice be maintained by infrastructure such as CI/CD, API security, logging, monitoring?**

IET is developing microservices to be compatible with the pipelines being developed by CST for `vets-api` in EKS. We are putting a focus on leveraging existing tools and pipelines instead of deploying new solutions (see https://vfs.atlassian.net/wiki/spaces/IET/pages/2109636630/ADR+Microservices+on+EKS). We have been working closely with infrastructure to ensure that we follow ‘standard’ processes. 

**What determines that something should be made a microservice?**

In Q1 IET developed a prioritization framework  https://vfs.atlassian.net/wiki/spaces/IET/pages/2054684691/Vets-api+Microservice+Framework. This document is a bit out of date, but goes into why file-upload had been prioritized for our next service (https://vfs.atlassian.net/wiki/spaces/IET/pages/2108620842/RFC+File+Upload+Selection). The services that we have developed thus far (Sidekiq UI and file upload) were given additional prioritization because we felt like they were good candidates for initial implementations. 

**Should we consider non-ruby (everything in VSP runs on ruby today) for the first of its kind in microservices? Or should we build a standard cookie cutter app first in Sinatra that allows for more broad support initially. I think creating something with the standard path is more ideal than start with an exception (python instead of ruby).**

This is the approach that was followed with Sidekiq UI (Sinatra app developed for more broad support — and also to give us an opportunity to develop integration with AMS in ruby). Sidekiq UI represents the simplest microservice, and is being developed in ruby for the reasons described. 

That being said, I think question here is whether or not we want to define ruby as being the standard path. In the long term (that is, when the platform has fully committed to a microservices architecture) I believe that it is important to allow teams to select the language best suited to the problem domain they are trying to solve. 

**How do we measure when something should be split out from the monolith? Is it strictly based on lack of rails engineers on the product team, or are their documented performance gains to be made?**

Is this question referring to the general question of service prioritization (figuring out which services we want to target) or the general desire to break `vets-api` into smaller pieces of functionality? 

**What is the plan/How are we going to measure if the microservice is meeting the goals of why the microservice was created instead of being in the monolith? What should we measure to determine this?**

Currently the goal of the microservices that we are implementing are to act as example implementations for future services, and as such the services that we have prioritized (file upload & sidekiqUI) were chosen more for their suitability for an example implementation of a platform microservice. 

The document I linked regarding choosing microservices includes some “things” we could look  into — but at this point it does not include concrete metrics we could measure (ie: we include “developer impact” as a parameter that we are considering, but do not actually define what metrics we can look at to answer this question). If we are choosing services based on their suitability as example/pioneer services, we need to define was this means (the creation of white papers? adrs? Etc?)   I think this is a question that requires more thought from my team before continuing forward — defining specifically what defines a successful service.

Reply:
> How do we measure when something should be split out from the monolith? Is it strictly based on lack of rails engineers on the product team, or are their documented performance gains to be made?
> 
> Is this question referring to the general question of service prioritization (figuring out which services we want to target) or the general desire to break vets-api into smaller pieces of functionality?

For this question, I want to better understand how we measure the services within the monolith to determine a) this service should be pulled out from the monolith, and b) how do we measure the service after its become a microservice to determine the hypothesis of the improvements are being met. With "b)" it could help influence future priorities around microservice development.